src/lib/when.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, reparentNodes, DirectiveFn, NodePart} from '../core.js';
+
+interface PartCache {
+  truePart: NodePart;
+  falsePart: NodePart;
+  prevCondition?: boolean;
+  cacheContainer: DocumentFragment;
+}
+
+const partCaches = new WeakMap<NodePart, PartCache>();
+
+/**
+ * Efficiently switches between two templates based on the given condition. The rendered
+ * content is cached, and re-used when switching conditions. Templates are evaluated
+ * lazily, so the passed values must be functions.
+ *
+ * While this directive can render any regular part, it makes the most sense when used
+ * with TemplateResulte since most other values are dirty checked already.
+ *
+ * Example:
+ *
+ * let checked = false;
+ *
+ * html`
+ *   when(checked, () => html`Checkmark is checked`, () => html`Checkmark is not checked`);
+ * `
+ *
+ * @param condition the condition to test against
+ * @param trueValue the value to render given a true condition
+ * @param falseValue the value to render given a false condition
+ */
+export const when = (condition: boolean, trueValue: () => any, falseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      const partCache = getPartCache(part);
+      const { truePart, falsePart, cacheContainer, prevCondition } = partCache;
+
+      const nextPart = condition ? truePart : falsePart;
+      const nextValue = condition ? trueValue() : falseValue();
+
+      // Swap nodes if condition changed from previous
+      if (condition !== prevCondition) {
```

Comment:
You are checking for equality against the `prevCondition`, but what you really want is to check if they are both truthy or both falsy. Does TypeScript automatically add a boolean conversion to `condition` because it is typed as a boolean? If not, it might be worth to manually cast `condition` to a boolean value with `condition = !!condition` to make sure this check returns what you expect.

Reply:
Typescript forces consumers to use a boolean value, which I like. But for JS I guess the expectation that it checks truthiness instead. If we want to support that, then it needs some extra checks as now I depend on boolean equality so that the initial value of undefined for the cached condition does not cause it to match the false condition.

Reply:
Don't assume that consumer will be using TypeScript. You can rely on the type as much as you want to assert that the docs have to be followed exactly. In this case, JS users are so used to "truthiness" that we should support it. 

`condition = !!condition` is a fine way to do this, and you could even type `condition` as `unknown`. Make sure you test initial render in both the true and false cases.